### PR TITLE
[receiver/mongodbatlasreceiver] add metrics project config

### PR DESCRIPTION
**Pull request by JonathanWamsley**
**Description:** <Describe what has changed.>

This feature adds a Project Config for the metrics to filter by Project name and or clusters.

**Link to tracking Issue:** <Issue number if applicable>
#28865 

**Testing:** <Describe what testing was performed and which tests were added.>
- Added test for cluster filtering
- Tested project name alone, project name with IncludeClusters and project name with ExcludeClusters on a live environment with success.

**Documentation:** <Describe the documentation added.>
Added optional project config fields to README

### CODE REVIEW

File: receiver/mongodbatlasreceiver/receiver.go [hunk 1]
```diff
@@ -77,46 +81,97 @@ func (s *mongodbatlasreceiver) shutdown(context.Context) error {
...
 	}
+
 	return nil
 }
 
+// shouldProcessCluster checks whether a given cluster should be processed based on the project configuration.
+func shouldProcessCluster(projectCfg *ProjectConfig, clusterName string) bool {
+	_, isIncluded := projectCfg.includesByClusterName[clusterName]
+	_, isExcluded := projectCfg.excludesByClusterName[clusterName]
+	return (isIncluded && !isExcluded) || (len(projectCfg.IncludeClusters) == 0 && !isExcluded)
```
Comment (dehaansa) [inline receiver/mongodbatlasreceiver/receiver.go:172]: Feels like the logic is a little unclear here, could read a bit clearer if we reorganize the logic to represent the cases we're evaluating.
```
(len(projectCfg.IncludeClusters) == 0 || isIncluded) && !isExcluded
```

File: receiver/mongodbatlasreceiver/receiver_test.go [hunk 2]
```diff
@@ -71,3 +71,64 @@ func TestTimeConstraints(t *testing.T) {
...
+				ExcludeClusters: []string{"Cluster2"},
+				excludesByClusterName: map[string]struct{}{
+					"Cluster2": {},
+				},
+			},
+			clusterName: "Cluster2",
+			want:        false,
+		},
+		{
+			name: "cluster not in include or exclude list",
```
Comment (dehaansa) [inline receiver/mongodbatlasreceiver/receiver_test.go:105]: Is this a valid case? Doesn't the config validation require that at max one of include/exclude lists be used?

File: receiver/mongodbatlasreceiver/receiver_test.go [hunk 3]
```diff
@@ -71,3 +71,64 @@ func TestTimeConstraints(t *testing.T) {
...
+
+func TestShouldProcessCluster(t *testing.T) {
+	tests := []struct {
+		name        string
+		projectCfg  *ProjectConfig
+		clusterName string
+		want        bool
+	}{
+		{
+			name: "included cluster",
```
Comment (dehaansa) [inline receiver/mongodbatlasreceiver/receiver_test.go:83]: Add a test case for include list, but cluster not in list, and exclude list, but cluster not in list.

File: receiver/mongodbatlasreceiver/config.go [hunk 4]
```diff
@@ -133,6 +134,12 @@ var (
 func (c *Config) Validate() error {
 	var errs error
 
+	for _, project := range c.Projects {
+		if len(project.ExcludeClusters) != 0 && len(project.IncludeClusters) != 0 {
```
Comment (fatsheep9146) [inline receiver/mongodbatlasreceiver/config.go:138]: could you add test for this line?

File: receiver/mongodbatlasreceiver/README.md [hunk 5]
```diff
@@ -42,6 +42,11 @@ MongoDB Atlas [Documentation](https://www.mongodb.com/docs/atlas/reference/api/l
 - `granularity` (default `PT1M` - See [MongoDB Atlas Documentation](https://docs.atlas.mongodb.com/reference/api/process-measurements/))
 - `collection_interval` (default `3m`) This receiver collects metrics on an interval. Valid time units are `ns`, `us` (or `µs`), `ms`, `s`, `m`, `h`.
 - `storage` (optional) The component ID of a storage extension which can be used when polling for `alerts` or `events` . The storage extension prevents duplication of data after a collector restart by remembering which data were previously collected.
+- `projects` (optional for metrics)
```
Comment (fatsheep9146) [inline receiver/mongodbatlasreceiver/README.md:45]: This description made me hard to know this field is a slice, could you optimize your description?

File: receiver/mongodbatlasreceiver/receiver.go (hunk 1 above)
Comment (JonathanWamsley) [reply receiver/mongodbatlasreceiver/receiver.go]: I split the return function into 2 parts with more comments to make it more clear.

File: receiver/mongodbatlasreceiver/receiver_test.go (hunk 2 above)
Comment (JonathanWamsley) [reply receiver/mongodbatlasreceiver/receiver_test.go]: I see, I updated the valid cases. I revised the test cases based on the config `Validate` function which when called asserts that there won't be an included and excluded cluster at the same time with:
```go
		if len(project.ExcludeClusters) != 0 && len(project.IncludeClusters) != 0 {
			errs = multierr.Append(errs, errClusterConfig)
		}
```

and `populateIncludesAndExcludes` function is called so I only need to put in the `IncludeClusters` or `ExcludeClusters` values:
```go
func (pc *ProjectConfig) populateIncludesAndExcludes() {
	pc.includesByClusterName = map[string]struct{}{}
	for _, inclusion := range pc.IncludeClusters {
		pc.includesByClusterName[inclusion] = struct{}{}
	}

	pc.excludesByClusterName = map[string]struct{}{}
	for _, exclusion := range pc.ExcludeClusters {
		pc.excludesByClusterName[exclusion] = struct{}{}
	}
}
```

File: receiver/mongodbatlasreceiver/receiver_test.go (hunk 3 above)
Comment (JonathanWamsley) [reply receiver/mongodbatlasreceiver/receiver_test.go]: The functions mentioned above should make it more clear that this combination is not possible.

File: receiver/mongodbatlasreceiver/config.go (hunk 4 above)
Comment (JonathanWamsley) [reply receiver/mongodbatlasreceiver/config.go:138]: Added tests

File: receiver/mongodbatlasreceiver/README.md (hunk 5 above)
Comment (JonathanWamsley) [reply receiver/mongodbatlasreceiver/README.md]: added a better description

**Comment by JonathanWamsley** [comment]
Hey @dehaansa  and @fatsheep9146 , PR feedback has been applied and it is ready to be looked at again! Thanks 🙏

**Review by dehaansa — APPROVED** [review]

**Review by fatsheep9146 — APPROVED** [review]

**Comment by fatsheep9146** [comment]
@djaglowski could you help review this pr?

File: receiver/mongodbatlasreceiver/receiver.go [hunk 6]
```diff
@@ -77,47 +81,106 @@ func (s *mongodbatlasreceiver) shutdown(context.Context) error {
...
-				}
+	return nil
+}
 
-				if err := s.extractProcessDiskMetrics(ctx, time, org.Name, project, process, clusterName, providerValues); err != nil {
-					return fmt.Errorf("error when polling process disk metrics from MongoDB Atlas for process %s: %w", process.ID, err)
-				}
-			}
+func (s *mongodbatlasreceiver) processProject(ctx context.Context, time timeconstraints, project *mongodbatlas.Project, projectCfg *ProjectConfig) error {
+	org, err := s.client.GetOrganization(ctx, project.OrgID)
```
Comment (djaglowski) [inline receiver/mongodbatlasreceiver/receiver.go:129]: We shouldn't have to do this again for projects which we obtained _from_ an organization i.e. the default situation.

File: receiver/mongodbatlasreceiver/receiver.go [hunk 7]
```diff
@@ -77,47 +81,106 @@ func (s *mongodbatlasreceiver) shutdown(context.Context) error {
...
+		_, isExcluded := projectCfg.excludesByClusterName[clusterName]
+		return !isExcluded
+	}
+
+	// If included clusters are specified, only process clusters that are included and not excluded.
+	_, isIncluded := projectCfg.includesByClusterName[clusterName]
+	_, isExcluded := projectCfg.excludesByClusterName[clusterName]
+	return isIncluded && !isExcluded
+}
+
```
Comment (djaglowski) [inline receiver/mongodbatlasreceiver/receiver.go:183]: Is this necessarily related to including/excluding projects? Can it be a separate PR?

File: receiver/mongodbatlasreceiver/receiver.go (hunk 6 above)
Comment (djaglowski) [reply receiver/mongodbatlasreceiver/receiver.go:129]: I suggest passing it in. We already have it when there are no projects lists. Otherwise, we have to look it up per project anyways.

File: receiver/mongodbatlasreceiver/receiver.go [hunk 8]
```diff
@@ -77,47 +81,106 @@ func (s *mongodbatlasreceiver) shutdown(context.Context) error {
...
+	var projectsWithConfig []ProjectWithConfig
+
+	if len(s.cfg.Projects) > 0 {
+		for _, projectCfg := range s.cfg.Projects {
+			project, err := s.client.GetProject(ctx, projectCfg.Name)
+			if err != nil {
+				s.log.Error("error retrieving project", zap.String("projectName", projectCfg.Name), zap.Error(err))
+				continue
+			}
+			projectsWithConfig = append(projectsWithConfig, ProjectWithConfig{Project: project, ProjectConfig: projectCfg})
```
Comment (djaglowski) [inline receiver/mongodbatlasreceiver/receiver.go:99]: Previously we did not need such a struct because we would process each project fully before moving on to the next. There may be a case for changing that pattern, but I think it needs to be made - OR, we just call `processProject` here instead, which would be a lot simpler codewise.

File: receiver/mongodbatlasreceiver/receiver.go [hunk 9]
```diff
@@ -77,47 +81,106 @@ func (s *mongodbatlasreceiver) shutdown(context.Context) error {
...
+				s.log.Error("error retrieving projects", zap.String("orgID", org.ID), zap.Error(err))
+				continue
 			}
-
-			processes, err := s.client.Processes(ctx, project.ID)
-			if err != nil {
-				return fmt.Errorf("error retrieving MongoDB Atlas processes for project %s: %w", project.ID, err)
+			for _, project := range proj {
+				// Since there is no specific ProjectConfig for these projects, pass nil.
+				projectsWithConfig = append(projectsWithConfig, ProjectWithConfig{Project: project, ProjectConfig: nil})
```
Comment (djaglowski) [inline receiver/mongodbatlasreceiver/receiver.go:114]: Call `processProject` here too instead.

Comment (JonathanWamsley) [reply receiver/mongodbatlasreceiver/receiver.go]: updated

File: receiver/mongodbatlasreceiver/receiver.go (hunk 8 above)
Comment (JonathanWamsley) [reply receiver/mongodbatlasreceiver/receiver.go]: updated calling the `processProject`

File: receiver/mongodbatlasreceiver/receiver.go (hunk 6 above)
Comment (JonathanWamsley) [reply receiver/mongodbatlasreceiver/receiver.go]: updated

File: receiver/mongodbatlasreceiver/receiver.go (hunk 7 above)
Comment (JonathanWamsley) [reply receiver/mongodbatlasreceiver/receiver.go]: Yes? This PR is adding a project config which is reusing an existing project config pattern that includes a project name, exclude and include clusters slice. I think it is reasonable for all the project config fields to be implemented if it is possible.

File: receiver/mongodbatlasreceiver/receiver.go [hunk 10]
```diff
@@ -78,46 +82,100 @@ func (s *mongodbatlasreceiver) shutdown(context.Context) error {
...
-		return fmt.Errorf("error retrieving organizations: %w", err)
-	}
-	for _, org := range orgs {
-		projects, err := s.client.Projects(ctx, org.ID)
-		if err != nil {
-			return fmt.Errorf("error retrieving projects: %w", err)
-		}
-		for _, project := range projects {
-			nodeClusterMap, providerMap, err := s.getNodeClusterNameMap(ctx, project.ID)
+	if len(s.cfg.Projects) > 0 {
```
Comment (djaglowski) [inline receiver/mongodbatlasreceiver/receiver.go:85]: Rather than if / else both wrapping large blocks of code, we should only need an `if` which returns at the end.

Similarly, these are complex enough that it would be helpful to move them into separate functions:

```
func (s *mongodbatlasreceiver) poll(ctx context.Context, time timeconstraints) error {
	if len(s.cfg.Projects) == 0 {
		return s.pollAllProjects(ctx, time)
	}
	return s.pollProjects(ctx, time)
}
```

Comment (JonathanWamsley) [reply receiver/mongodbatlasreceiver/receiver.go]: updated code where I could to make it look better.

File: receiver/mongodbatlasreceiver/receiver.go [hunk 11]
```diff
@@ -77,47 +81,113 @@ func (s *mongodbatlasreceiver) shutdown(context.Context) error {
 	return s.client.Shutdown()
 }
 
+// poll decides whether to poll all projects or a specific project based on the configuration.
 func (s *mongodbatlasreceiver) poll(ctx context.Context, time timeconstraints) error {
+	if len(s.cfg.Projects) == 0 {
+		return s.pollAllProjects(ctx, time)
+	}
+	return s.pollProject(ctx, time)
```
Comment (djaglowski) [inline receiver/mongodbatlasreceiver/receiver.go:89]:
```suggestion
	return s.pollProjects(ctx, time)
```

File: receiver/mongodbatlasreceiver/receiver.go [hunk 12]
```diff
@@ -77,47 +81,113 @@ func (s *mongodbatlasreceiver) shutdown(context.Context) error {
...
 
-			processes, err := s.client.Processes(ctx, project.ID)
-			if err != nil {
-				return fmt.Errorf("error retrieving MongoDB Atlas processes for project %s: %w", project.ID, err)
-			}
-			for _, process := range processes {
-				clusterName := nodeClusterMap[process.UserAlias]
-				providerValues := providerMap[clusterName]
+// pollProject handles polling for specific projects as configured.
+func (s *mongodbatlasreceiver) pollProject(ctx context.Context, time timeconstraints) error {
```
Comment (djaglowski) [inline receiver/mongodbatlasreceiver/receiver.go:115]:
```suggestion
func (s *mongodbatlasreceiver) pollProjects(ctx context.Context, time timeconstraints) error {
```

**Review by djaglowski — APPROVED** [review]
